File: sec-blueprints/service-to-service/api-onbehalfof-springsecurity/src/test/java/no/nav/security/examples/springsecurity/client/DownstreamResourceClientTest.java
package no.nav.security.examples.springsecurity.client;

import no.nav.security.examples.springsecurity.config.OAuth2ClientConfiguration;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.security.oauth2.client.OAuth2AuthorizeRequest;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientManager;
import org.springframework.security.oauth2.core.OAuth2AccessToken;
import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse;
import org.springframework.security.test.context.support.WithAnonymousUser;
import org.springframework.web.client.RestTemplate;

import java.io.IOException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@SpringBootTest(classes = OAuth2ClientConfiguration.class)
class DownstreamResourceClientTest {

    private static final String TOKEN = "tokenvalue";

    @MockBean
    private OAuth2AuthorizedClientManager authorizedClientManager;
    @Mock
    private OAuth2AuthorizedClient authorizedClient;

    @Autowired
    private RestTemplate restTemplate;

    private DownstreamResourceClient downstreamResourceClient;
    private MockWebServer server;

    @BeforeEach
    void setup() throws IOException {
        MockitoAnnotations.initMocks(this);
        this.server = new MockWebServer();
        this.server.start();
        when(authorizedClientManager.authorize(any(OAuth2AuthorizeRequest.class)))
                .thenReturn(authorizedClient);
        when(authorizedClient.getAccessToken()).thenReturn(OAuth2AccessTokenResponse
                .withToken(TOKEN)
                .tokenType(OAuth2AccessToken.TokenType.BEARER)
                .build().getAccessToken());

        downstreamResourceClient = new DownstreamResourceClient(
                "http://localhost:" + server.getPort(), restTemplate);
    }

    @AfterEach
    void teardown() throws IOException {
        server.shutdown();
    }

    @Test
    @WithAnonymousUser
    void testPingWithClientCredentials() throws InterruptedException {
        server.enqueue(new MockResponse().setBody("pong"));
        assertThat(downstreamResourceClient.ping()).isNotBlank();
        RecordedRequest recordedRequest = this.server.takeRequest();
        assertThat(recordedRequest.getHeader("Authorization")).contains(TOKEN);
    }

}

File: sec-blueprints/service-to-service/api-onbehalfof-ktor/src/test/kotlin/no/nav/security/examples/DownstreamResourceClientTest.kt
package no.nav.security.examples

import com.github.michaelbull.result.Ok
import com.github.michaelbull.result.expect
import com.github.michaelbull.result.expectError
import io.ktor.client.features.ClientRequestException
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.mockk.coEvery
import io.mockk.mockk
import kotlinx.coroutines.runBlocking
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

private val validAccessToken = AccessToken(
    accessToken = "validAccessToken",
    expiresIn = 3600,
    tokenType = "bearer"
)
private const val downstreamClientId: String = "clientId"

class DownstreamResourceClientTest {

    private val downstreamMockServer: MockWebServer = MockWebServer()
    private val azureAdClientMock: AzureAdClient = mockk()
    private val downstreamResourceClient: DownstreamResourceClient =
        DownstreamResourceClient(azureAdClient = azureAdClientMock)
    private lateinit var downstreamUrl: String

    @BeforeEach
    fun setup() {
        downstreamMockServer.start()
        downstreamUrl = "http://localhost:${downstreamMockServer.port}/ping"
    }

    @AfterEach
    fun teardown() {
        downstreamMockServer.shutdown()
    }

    @Test
    fun `given a valid access token, fetching resource from downstream resource should succeed`() {
        downstreamMockServer.enqueue(
            MockResponse().setBody(
                """
                    {
                        "ping" : "You've reached a protected endpoint at downstream API"
                    }
                """.trimIndent()
            ).setHeader(HttpHeaders.ContentType, ContentType.Application.Json.toString())
        )
        coEvery { azureAdClientMock.getOnBehalfOfAccessTokenForResource(any(), any()) } returns Ok(
            validAccessToken
        )
        runBlocking {
            downstreamResourceClient.get(
                Resource(downstreamClientId, downstreamUrl),
                accessToken = validAccessToken.accessToken
            )
        }.expect { "valid response from downstream server" }
            .let { resource: Resource ->
                assertTrue(resource.response.toString().isNotEmpty())
                assertTrue(downstreamMockServer.takeRequest().getHeader("Authorization")!!.contains(validAccessToken.accessToken))
            }
    }

    @Test
    fun `given an invalid access token, fetching resource from downstream resource should fail`() {
        downstreamMockServer.enqueue(MockResponse().setResponseCode(401).setBody("invalid access token"))
        val invalidOboAccessToken = "invalidOboAccessToken"
        val invalidAccessToken = "invalidAccessToken"
        coEvery { azureAdClientMock.getOnBehalfOfAccessTokenForResource(any(), any()) } returns Ok(
            validAccessToken.copy(accessToken = invalidOboAccessToken)
        )
        runBlocking {
            downstreamResourceClient.get(
                Resource(downstreamClientId, downstreamUrl),
                accessToken = invalidAccessToken
            )
        }.expectError { "unauthorized response from downstream server" }
            .let { error ->
                val exception = error.throwable as ClientRequestException
                assertEquals(exception.response?.status, HttpStatusCode.Unauthorized)
                assertTrue(downstreamMockServer.takeRequest().getHeader("Authorization")!!.contains(invalidOboAccessToken))
            }
    }
}
